docs: Add basic logging documentation (#926)
DiamondLightSource/blueapi: docs/how-to/configure-logging.md

# Configure Logging

By default BlueAPI will log to stdout at the [INFO level](https://docs.python.org/3/library/logging.html#logging-levels), but can be reconfigured to log at any level, and to output to Graylog.

> As the BlueAPI log handlers are set at root level they will recieve messages from all loggers in the same process that [allow propagation](https://docs.python.org/3/library/logging.html#logging.Logger.propagate).

Reviewer: Nit: The allowing propagation stuff seems like a lot of detail. I didn't even know you could turn this off, I suspect those new to Python logging will never need to touch it.

Author: References to propagation removed

> To add logging to a custom module in BlueAPI, instantiate a logger from the [standard library logging package, then use any of its log methods](https://docs.python.org/3/library/logging.html#logger-objects). Propagation is enabled by default.

Reviewer: Could: A small 2-3 line example here would be great.

Author: Added (very) basic example

# Configure Logging

By default BlueAPI will log to stdout at the [INFO level](https://docs.python.org/3/library/logging.html#logging-levels), but can be reconfigured to log at any level, and to output to Graylog.

When logging to [Graylog](https://graylog.org) is enabled, BlueAPI will also continue to log to stdout.

# BlueAPI Cofiguration

:::{seealso}
[Configure the Application](./configure-app.md)
:::

An example logging config is shown below:
```
  logging:
    level: "INFO"
    graylog:
      enabled: True
      host: "graylog-log-target.diamond.ac.uk"
      port: 12232
```

> Here, BlueAPI will accept all messages are INFO level and above and will forward these to both the referenced Graylog address, and to stdout.

Author: Agreed - the sentence has been reworded.

# Instrumenting Libraries
To instrument to a custom module in BlueAPI, instantiate a logger from the [standard library logging package, then use any of its log methods](https://docs.python.org/3/library/logging.html#logger-objects).

> BlueAPI is written to intercept logs from any python code it executes.

Reviewer 2: ```suggestion
BlueAPI is configured to handle logging from any python code it executes.
```

This is a bit picky but I'm not aware of blueapi doing any logging handling/intercepting itself.

```
import logging
logger = logging.getLogger(__name__)
logger.info("FOO")
```

# Kubernetes

Services hosted on the DLS clusters automatically have their stdout forwarded to Graylog via a service called fluentd. Due to this, BlueAPI services hosted on the cluster will always log to Graylog.

> In this instance, the difference between enabling and disabling graylog in BlueAPI's configuration, is that fluentd only fowards plaintext logs, while BlueAPI's native graylog produces structured data with rich metadata included.

Reviewer 2: Presumably the logs will be duplicated if graylog is configured directly. Is it worth mentioning that and/or is it possible to disable stdout logs while keeping graylog?

Author: This is a good point - we have set up the helm chart to automatically disable fluentd when direct graylog logging is enabled. That is:
```
      annotations:
      {{- if ((.Values.worker.logging).graylog).enabled }}
        fluentd-ignore: "true"
      {{- end }}
```

I'll add a sentence explaining this.

# Where to Find Logs

By default logs can be found wherever your stdout is. 

> If Graylog is enabled, logs will be forwarded to whichever Graylog instance the configuration addresses. By default this is the main Diamond instance, which can be accessed via [graylog.diamond.ac.uk](https://graylog.diamond.ac.uk/)

Reviewer 2: where is the default being set? In the helm charts?

Author: When [set_up_logging](https://github.com/DiamondLightSource/blueapi/blob/e0a51d5ed32923dda5c6390ee9b774d5a4b92259/src/blueapi/log.py#L24) is called it always sets up a stream handler which will forward all logs to stdout.

If Graylog is enabled, both a Graylog handler and a stream handler are set up.

Currently there is no case where the logs are not sent to stdout. Would this be useful?

Reviewer 2: I meant the `By default this is the main Diamond instance...` bit. Where is diamond's graylog the default?

Reviewer 2: Fair enough, the docs match reality but it's a bit odd for Diamond specific graylog config to be in the default values when other urls use example values, eg `example.diamond.ac.uk` or `http://opentelemetry-collector.tracing`. Probably not an issue for this PR though

Reviewer 3: It would alarm me more if it was also enabled by default, so could be worse. I would say it's mainly a "prettiness" thing.

> If your BlueAPI server is running on the cluster, stdout is likely being forwarded to the above Graylog instance.

Reviewer 2: ```suggestion
If your BlueAPI server is running on the Diamond cluster, stdout is likely being forwarded to the above Graylog instance.
```

# Configure Logging

By default BlueAPI will log to stdout at the [INFO level](https://docs.python.org/3/library/logging.html#logging-levels), but can be reconfigured to log at any level, and to output to Graylog.

> When Graylog is enabled, BlueAPI will continue to log to stdout.

Reviewer 3: ```suggestion
When logging to [Graylog](https://graylog.org) is enabled, BlueAPI will also continue to log to stdout.
```

# BlueAPI Cofiguration

:::{seealso}
[Configure the Application](./configure-app.md)
:::

An example logging config is shown below:
```

> worker:
>     logging:
>       level: "INFO"
>       graylog:
>         enabled: True
>         host: "graylog-log-target.diamond.ac.uk"
>         port: 12232

Reviewer 3: The `worker` part is helm-specific

```suggestion
  logging:
    level: "INFO"
    graylog:
      enabled: True
      host: "graylog-log-target.diamond.ac.uk"
      port: 12232
```

```

Here, BlueAPI will accept all messages are INFO level and above and will forward these to both the referenced Graylog address, and to stdout.

# Instrumenting Libraries
To instrument to a custom module in BlueAPI, instantiate a logger from the [standard library logging package, then use any of its log methods](https://docs.python.org/3/library/logging.html#logger-objects).

BlueAPI is written to intercept logs from any python code it executes.

```
import logging
logger = logging.getLogger(__name__)
logger.info("FOO")
```

> # Running BlueAPI on the Cluster

Reviewer 3: ```suggestion
# Kubernetes
```

> Services hosted on the cluster automatically have their stdout forwarded to Graylog via a service called fluentd. Due to this, BlueAPI services hosted on the cluster will always log to Graylog.

Reviewer 3: ```suggestion
Services hosted on the DLS clusters automatically have their stdout forwarded to Graylog via a service called fluentd. Due to this, BlueAPI services hosted on the cluster will always log to Graylog.
```

In this instance, the difference between enabling and disabling graylog in BlueAPI's configuration, is that fluentd only fowards plaintext logs, while BlueAPI's native graylog produces structured data with rich metadata included.

> # Were to Find Logs

Reviewer 3: ```suggestion
# Where to Find Logs
```

# Configure Logging

> By default BlueAPI will log to stdout at the [INFO level](https://docs.python.org/3/library/logging.html#logging-levels), but can be reconfigured to log at any level, and to output to Graylog.

Reviewer 4: ```suggestion
By default BlueAPI will log to stdout at the [INFO level](https://docs.python.org/3/library/logging.html#logging-levels), but can be configured to log at any level, and to output simultaneously at the same level to [Graylog](https://graylog.org).
```

> 
> When logging to [Graylog](https://graylog.org) is enabled, BlueAPI will also continue to log to stdout.
> 

Reviewer 4: ```suggestion

```

# BlueAPI Cofiguration

:::{seealso}
[Configure the Application](./configure-app.md)
:::

An example logging config is shown below:
```yaml
  logging:
    level: "INFO"
    graylog:
      enabled: True
      host: "graylog-log-target.diamond.ac.uk"

>       port: 12232

Reviewer 4: ```suggestion
      port: 12231
```

We should adjust the defaults to hit the right graylog too

```

With this configuration, all logs at INFO level or above will be logged to both stdout and the configured graylog.

# Instrumenting Libraries

> To instrument to a custom module in BlueAPI, instantiate a logger from the [standard library logging package, then use any of its log methods](https://docs.python.org/3/library/logging.html#logger-objects).
> 
> BlueAPI is configured to handle logging from any python code it executes.

Reviewer 4: ```suggestion
Logs from plans run by BlueAPI will use inherit the configured logging, simply instantiate a logger from the [standard library logging package, then use any of its log methods](https://docs.python.org/3/library/logging.html#logger-objects).
```

```

> import logging
> logger = logging.getLogger(__name__)
> logger.info("FOO")

Reviewer 4: ```suggestion
import logging
logger = logging.getLogger(__name__)

def my_plan() -> MsgGenerator:
  logger.info("Running my_plan")
  
logger.debug("Loaded module with my_plan")
```

```

# Kubernetes

> Services hosted on the DLS clusters automatically have their stdout forwarded to Graylog via a service called fluentd. Due to this, BlueAPI services hosted on the cluster will always log to Graylog.
> 
> When BlueAPI's native Graylog support is enabled it forwards structured data rather than plaintext.
> 
> When BlueAPI's native Graylog is enabled fluentd will be automatically disabled to avoid log duplication.

Reviewer 4: ```suggestion
Services hosted on the DLS clusters automatically have their stdout forwarded to Graylog via a service called fluentd. Therefore BlueAPI hosted on the cluster will always have logs available from Graylog, however BlueAPI's native Graylog forwards structured data rather than plaintext.

When native Graylog is enabled fluentd will be automatically disabled to avoid log duplication.
```
